<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta name="generator" content="HTML Tidy for Windows (vers 14 February 2006), see www.w3.org" />
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
    <title>
      Affymetrix Developer Network
    </title>
    <link href="css/styles.css" rel="stylesheet" "text/css" />
  </head>
  <body>
    <div id="wrapper">
      <div id="page">
        <div id="header">
          <img alt="Affymetrix.com" src="images/header_banner.jpg" />
        </div>
        <div id="content">
<h1>Reading Command Console Data Files</h1>

 
		<h2>INTRODUCTION</h2>
	

	<p>The Fusion SDK provides source code for reading both GCOS 1.x and Command Console data files. The way the SDK is organized is depicted in the diagram below.</p>
	<p><img border="0" src="fusion_layers.gif"></p>

<p>The GCOS files parsers are file specific as each data file has its own format.</p>

<p>The Command Console (CC) Array file is an XML file. The XML parser used is the Apache Xerces software.</p>

<p>The CC data files use the same flexible file format. This format contains information to describe the contents (#columns, data type in each column, etc.). A set of classes provide the ability to read any data file of this format, regardless of file extension. For existing files such as the CEL and CHP files which contain results generated by the current GCOS/GTYPE/GDAS software, file specific interpreters provide higher level interfaces. These interpreters make assumptions about the number of data groups, sets and column which the CC software will conform to. For assays that will write CHP files with newer data (such as the Tiling and ExACT software) the fusion layer will utilize the generic file parsers.</p>

<p>The information below will describe how to read each data file type using the Fusion APIs.</p>
		<h2>ARRAY FILE</h2>
	

	<p>The CC array file contains information about a GeneChip probe array and end user attributes.</p>

<p>The top level Fusion objects to use are <b><i>FusionArrayFileReader</b></i> and <b><i>ArrayData</b></i>. These classes will parse CC array files, MAGE-ML files generated by the DTT 1.1 software and EXP files.</p>

<p>The workflow for reading the array file is:</p>

   <p>1. Create a FusionArrayFileReader object.<br />
   2. Read the file using the Read() function passing in the full path name of the file and the ArrayData object.</p>

	<h2>CEL FILE</h2>
	

	<p>The CEL file contains intensity data organized by X/Y feature coordinate. Included in the file is the name of the algorithm and the parameters used to create the file.</p>

<p>The top level Fusion object to use is <b><i>FusionCELData</b></i>.</p>

<p>The workflow for reading a CEL file is:</p>

   <p>1. Create a FusionCELData object.<br />
   2. Set the full path name of the input CEL file using the SetFileName() function.<br />
   3. Read the file using the Read() function. A false boolean value may be passed into the function if mask and outlier data are not to be read.<br />
   4. Use the GetParameters() function to retrieve the list of algorithm parameters.<br />
   5. Use the GetAlg() function to retrieve the name of the algorithm used to create the CEL file.<br />
   6. Various function are available to retrieve the intensity data. These include GetEntry() to retrieve the intensity, standard deviation and pixel count for a given feature and GetIntensity() to retrieve just the intensity value.</p>

<p>Note: Use the CDF file to organize the data by probe set.</p>
		
		<h2>CHP FILE</h2>
	

	<p>The CHP file contains the results from analyze intensity data in a CEL file. For expression arrays this is the probe set results such as signal and detection. For genotyping this is the allele calls and confidence scores.</p>

<p>Reading a CHP file involves several Fusion classes. To read the file, use the <b><i>FusionCHPData</b></i> and <b><i>FusionCHPDataReg</b></i> classes.</p>

         <blockquote>
		 <p><i>FusionCHPData *chp = FusionCHPDataReg::Read(fileName)</i>;  // fileName is the name of the input CHP file.</p>
		 </blockquote>

<p>To support the multitude of expected CHP file types, the CHP parser has been written using self registering classes with each self registered class providing the parsing capabilities for the specific CHP file type. The existing CHP parsers include:</p>

    <ul>
	<li>FusionCHPLegacyData - This class provides parsing for CHP files containing results from our legacy assays (expression, genotyping, resequencing, tag).</li>
    <li>FusionCHPTilingData - This class provides parsing for CHP files containing results from the Tiling Analysis Software.</li>
    <li>FusionCHPGenericData - This class provides a catch all for CHP files where a data specific interpreter is not provided. This class directly accesses the generic data file parser and data objects, no assumptions are made as to how many data groups, sets and columns are stored in the file.</li>
    <li>FusionCHPQuantificationData - This class provides parsing for CHP files containing results from the RMA or PLIER algorithms from the Expression Console software for 3' Expression arrays.</li>
    <li>FusionCHPQuantificationDetectionData - This class provides parsing for CHP files containing results from the Expression Console software for Exon arrays.</li>
    <li>FusionCHPMultiDataData - This class provides parsing for CHP files containing genotyping analysis results from the BRLMM Analysis Tool Version 2 or Genotyping Console&#0153 software.</li>
	</ul>

<p>Each of the above classes contains a function (FromBase) to convert the FusionCHPData object. An example of converting to the legacy class is:</p>

    <p><i>FusionCHPLegacyData *legacychp = FusionCHPLegacyData::FromBase(chp)</i>;</p>

<p>The return object is NULL if the file is not an expression, genotyping, resequencing or tag CHP file (those data types supported by the FusionCHPLegacyData object).</p>

<p>At this point you need to decide what to do. If your application reads expression data and you want to extract the signal values for each probe set then the workflow steps are:</p>

   <p>1. Verify that the returned pointer from the FromBase() function is not NULL.<br />
   2. Verify that the data in the CHP file is expression using the GetHeader().GetAssayType() function. The value should be "FusionExpression".<br />
   3. Loop over the number of probe sets - defined by the GetHeader().GetNumProbeSets() function.<br />
   4. Retrieve the probe set object using the GetExpressionResults() function. This returns a FusionExpressionProbeSetResults object.<br />
   5. The signal value is retrieved using the FusionExpressionProbeSetResults.GetSignal() function.</p>

<p>Note: As of GCOS 1.0 the CEL intensities are no stored in the CHP file. Use the CEL parser to retrieve intensity information.</p>
		
		<h2>CDF FILE</h2>
	

	<p>The CDF file provides a description of the array in terms of function of the probes.</p>

<p>The top level Fusion object to use is <b><i>FusionCDFData</b></i>.</p>

<p>The workflow for reading a CDF file is:</p>

   <p>1. Create a FusionCDFData object.<br />
   2. Set the full path name of the input CDF file using the SetFileName() function. The name of the file is the array type + ".CDF". The array type may be extracted from a CEL file using the FusionCELData.GetChipType() function.<br />
   3. Read the file using the Read() function.<br />
   4. The name of a probe set may be extracted using the GetProbeSetName() function.<br />
   5. Loop over the number of probe sets. The number of probe sets is determined using the GetHeader().GetNumProbeSets() function. The probe set information is retrieved in a FusionCDFProbeSetInformation object.<br />
   6. The probe set contains one or more groups (1 for expression, typically 2 or 4 for genotyping). The number of probe groups is determined using the GetNumGroups() function. Call the GetGroupInformation() function to retrieve the group information.<br />
   7. The group contains one or more probes, defined by the GetNumCells() function. Each probe information may be extracted using the GetCell() function.

<p>Once you have a probe object you can determine if it is a PM or MM probe by examining the PBase and TBase. PM probes have PBase equal to the compliment of TBase, otherwise it is a MM probe.</p>
        </div>
        <div class="clr"></div>
        <div id="footer">
          &#169; 2009 Affymetrix, Inc. All rights reserved.
        </div>
      </div>
    </div>
  </body>
</html>